File: android/app/src/main/java/com/noop/data/PairedDevice.kt
package com.noop.data

import androidx.room.Entity
import androidx.room.PrimaryKey

/*
 * Device-registry schema — the Android port of the Swift foundation in
 * Packages/WhoopStore (PairedDevice.swift + DeviceRegistryStore.swift) and the Database.swift v15
 * `pairedDevice` / `dayOwnership` migration. Added to Room as the v7 → v8 additive migration.
 *
 * `id` is the SAME string used as `deviceId` in every sample table's (deviceId, ts) key, so a
 * device's raw samples are already isolated by id — no per-row source column is needed. The existing
 * WHOOP keeps id "my-whoop" (zero sample-row migration); it is seeded `active` by MIGRATION_7_8.
 */

/**
 * A device the user has paired. PK = [id] (== `deviceId` in the sample tables). [capabilities] is the
 * comma-joined set of [Metric] rawValues, byte-for-byte matching the Swift store's encoding so a
 * cross-platform backup round-trips. Exactly one row is `active` at a time (invariant I1, enforced in
 * [DeviceRegistry.setActive]).
 */
@Entity(tableName = "pairedDevice")
data class PairedDeviceRow(
    @PrimaryKey
    val id: String,
    val brand: String,
    val model: String,
    val nickname: String?,
    /**
     * The strap's stable BLE peripheral identifier — on Android the [android.bluetooth.BluetoothDevice]
     * MAC address (also exposed as [com.noop.ble.WhoopBleClient.lastDeviceAddress]). Added by the v8 → v9
     * additive migration (the Android twin of the Swift `peripheralId` column). Lets the BLE client pin a
     * connect to ONE specific strap (multi-WHOOP) and lets a freshly-paired device be looked up by its
     * address. Nullable, no SQL DEFAULT (matches Room's generated column for a `String?` field); the
     * seeded "my-whoop" row keeps it NULL until the strap is (re)paired.
     */
    val peripheralId: String? = null,
    val sourceKind: String,
    val capabilities: String, // comma-joined Metric rawValues, e.g. "hr,hrv,sleep"
    val status: String,
    val addedAt: Long, // unix seconds
    val lastSeenAt: Long, // unix seconds
)

/**
 * Override of which device owns a given local day's displayed/scored metrics (invariant I2: a day's
 * scores are never blended across sources). PK = [day] ("YYYY-MM-DD"). [locked] = true means an
 * explicit decision (import-overlap resolution / user choice) that the resolver must honour over its
 * priority default. Port of the Swift `dayOwnership` table.
 */
@Entity(tableName = "dayOwnership")
data class DayOwnershipRow(
    @PrimaryKey
    val day: String,
    val deviceId: String,
    val locked: Boolean = false,
)

/** Lifecycle of a paired device. Stored as the lowercase enum name (Swift `DeviceStatus` rawValue). */
enum class DeviceStatus { active, paired, archived }

/** How a device's data reaches the store. Stored as the enum name (Swift `SourceKind` rawValue).
 *  [ftms] = a live FTMS gym machine (treadmill / indoor bike / rower / cross-trainer).
 *  [huami] = an EXPERIMENTAL Huami-family live HR source (Amazfit / Zepp incl. Helio, Xiaomi Mi Band):
 *  standard 0x180D when exposed, else the documented Huami custom HR characteristic, else an honest
 *  "needs pairing" message. (Garmin uses [liveBLE] — its live HR is the standard broadcast-HR path.)
 *  [oura] = an EXPERIMENTAL Oura ring live BLE source. Owns its OWN scanner/GATT (never touches the
 *  WHOOP client); decodes the ring's own raw signals + open HRV/sleep-phase tags and runs NOOP's own
 *  scoring, and surfaces an honest "needs pairing" state when the install key is absent (never Oura's
 *  encrypted readiness/sleep scores). Carried on string rawValue "oura"; no DB migration (the column is
 *  free-text and existing rows never carry it).
 *  Additive: existing rows never carry [ftms]/[huami]/[oura]; only the respective wizard paths write them. */
// `activityFile` (#137): a GPX/TCX/FIT activity file under the `activity-file` device. Distinct from
// `fileImport` (a whole-day WHOOP CSV export) so the day-owner resolver ranks it BELOW day-spanning
// imports — a 90-minute ride must never displace a full-day source with HR for the same day. It owns a
// day only when nothing else has data (a strap-less day). Additive: only the activity-file importer writes it.
enum class SourceKind { liveBLE, historyBLE, cloudImport, fileImport, ftms, huami, oura, activityFile }

/** A canonical metric a source can provide — drives capability-aware UI + the day-owner resolver.
 *  Stored as the enum name (Swift `Metric` rawValue) inside the comma-joined `capabilities` string. */
enum class Metric { hr, hrv, spo2, skinTemp, steps, sleep, strainLoad }

File: Packages/WhoopStore/Sources/WhoopStore/PairedDevice.swift
import Foundation

/// A device the user has paired. `id` is the same string used as `deviceId` in every sample table's
/// `(deviceId, ts)` key — so a device's raw samples are already isolated by id, with no per-row source
/// column needed. The existing WHOOP keeps id "my-whoop" (no sample migration).
public struct PairedDevice: Equatable, Sendable, Identifiable {
    public let id: String                 // == deviceId in sample tables; e.g. "my-whoop", "polar-h10-1A2B"
    public var brand: String              // "WHOOP", "Polar", "Garmin", "Oura"
    public var model: String              // "WHOOP 4.0", "H10", "Forerunner 265", "Oura (import)"
    public var nickname: String?          // user-renamable; nil → show brand+model
    public var peripheralId: String?      // CBPeripheral.identifier.uuidString (iOS/Mac); nil until adopted
    public var sourceKind: SourceKind
    public var capabilities: Set<Metric>
    public var status: DeviceStatus
    public var addedAt: Int               // unix seconds
    public var lastSeenAt: Int

    public init(id: String, brand: String, model: String, nickname: String? = nil,
                peripheralId: String? = nil,
                sourceKind: SourceKind, capabilities: Set<Metric>, status: DeviceStatus,
                addedAt: Int, lastSeenAt: Int) {
        self.id = id; self.brand = brand; self.model = model; self.nickname = nickname
        self.peripheralId = peripheralId
        self.sourceKind = sourceKind; self.capabilities = capabilities; self.status = status
        self.addedAt = addedAt; self.lastSeenAt = lastSeenAt
    }

    /// A user nickname wins; otherwise "Brand Model" — but collapse to just the model when it already
    /// carries the brand (so a WHOOP whose model is also "WHOOP" reads "WHOOP", not "WHOOP WHOOP", and a
    /// future "WHOOP 4.0" model reads "WHOOP 4.0").
    public var displayName: String {
        if let nickname { return nickname }
        if model.isEmpty || model == brand { return brand }
        if model.localizedCaseInsensitiveContains(brand) { return model }
        return "\(brand) \(model)"
    }

    /// True for a data-partition source (a cloud API pull or a file/CSV import) rather than a live,
    /// connectable device. UI that offers "make active" must exclude these: activating one would demote
    /// whatever live device (WHOOP, a BLE strap, Apple Watch) currently drives BLE routing + day-owner
    /// priority 0 — an import source has no live connection to hand that role to.
    public var isImportSource: Bool { sourceKind == .cloudImport || sourceKind == .fileImport }
}

public enum DeviceStatus: String, Sendable, CaseIterable { case active, paired, archived }

public enum SourceKind: String, Sendable, CaseIterable {
    case liveBLE, historyBLE, cloudImport, fileImport
    /// A live FTMS (Fitness Machine Service, 0x1826) gym machine — treadmill / indoor bike / rower /
    /// cross-trainer. Streams a live machine-data + HR session that records via the existing live-workout
    /// path. Additive: existing rows never carry this; only the gym-equipment wizard writes it.
    case ftms
    /// An EXPERIMENTAL Huami-family live HR source — Amazfit / Zepp (incl. Helio) and Xiaomi Mi Band.
    /// Reads the standard 0x180D HR service when exposed, else the documented Huami custom HR
    /// characteristic, else surfaces an honest "needs pairing we can't do" message. Best-effort, shipped
    /// behind the experimental add-device tier. Additive: only the experimental wizard writes it.
    /// (Garmin uses `liveBLE` — its live HR is the standard broadcast-HR path, no proprietary protocol.)
    case huami
    /// Apple Watch streamed via HealthKit (live HealthKit observer + background delivery). Apple-only,
    /// no Android twin. Additive: only the Apple Watch device registration writes it.
    case liveAppleWatch
    /// An EXPERIMENTAL live Oura Ring (gen 3/4/5) source over the ring's own clean-room BLE protocol
    /// (OuraProtocol package). Owns its OWN central/GATT, never the WHOOP path. Surfaces only the ring's
    /// decoded raw signals + open HRV/sleep-phase tags (capabilities {hr, hrv, spo2, skinTemp, sleep});
    /// NOOP computes its own Charge/Rest and never reads Oura's encrypted readiness/sleep scores. When a
    /// signal can't be read it stays "-" (Huami precedent), never faked. Additive (no DB migration): only
    /// the experimental add-device wizard's Oura path writes it.
    case oura
    /// A GPX/TCX/FIT activity file imported under the `activity-file` device (#137). Distinct from
    /// `fileImport` (a whole-day WHOOP CSV export) so the day-owner resolver can rank it BELOW day-spanning
    /// imports: a 90-minute ride must never displace a full-day source that has HR for the same day. It
    /// wins ownership only when nothing else has data (a genuinely strap-less day), where its measured HR
    /// lights the day Effort. Additive (no DB migration): only the activity-file importer writes it.
    case activityFile
}

/// Canonical metric a source can provide. Drives capability-aware UI + the day-owner resolver.
public enum Metric: String, Sendable, CaseIterable, Codable {
    case hr, hrv, spo2, skinTemp, steps, sleep, strainLoad
}
